src/lib/operators/table_scan/sorted_segment_search.hpp

```diff
@@ -0,0 +1,152 @@
+   * behind the last matching one.
+   */
+  IteratorType _get_first_bound() const {
+    if (_is_ascending) {
+      return std::lower_bound(_begin, _end, _search_value, [](const auto& segment_position, const auto& search_value) {
+        return segment_position.value() < search_value;
+      });
+    } else {
+      return std::lower_bound(_begin, _end, _search_value, [](const auto& segment_position, const auto& search_value) {
+        return segment_position.value() > search_value;
+      });
+    }
+  }
+
+  IteratorType _get_last_bound() const {
+    if (_is_ascending) {
+      return std::upper_bound(_begin, _end, _search_value, [](const auto& search_value, const auto& segment_position) {
+        return segment_position.value() > search_value;
+      });
+    } else {
+      return std::upper_bound(_begin, _end, _search_value, [](const auto& search_value, const auto& segment_position) {
+        return segment_position.value() < search_value;
+      });
+    }
+  }
+
+  // This function sets the offset(s) which delimit the result set based on the predicate condition and the sort order
+  void _set_begin_and_end() {
+    if (_predicate_condition == PredicateCondition::Equals) {
+      _begin = _get_first_bound();
+      _end = _get_last_bound();
+      return;
+    }
+
+    // clang-format off
+    if (_is_ascending) {
+      switch (_predicate_condition) {
+        case PredicateCondition::GreaterThanEquals: _begin = _get_first_bound(); return;
+        case PredicateCondition::GreaterThan: _begin = _get_last_bound(); return;
+        case PredicateCondition::LessThanEquals: _end = _get_last_bound(); return;
+        case PredicateCondition::LessThan: _end = _get_first_bound(); return;
+        default: Fail("Unsupported predicate condition encountered");
+      }
+    } else {
+      switch (_predicate_condition) {
+        case PredicateCondition::LessThanEquals: _begin = _get_first_bound(); return;
+        case PredicateCondition::LessThan: _begin = _get_last_bound(); return;
+        case PredicateCondition::GreaterThanEquals: _end = _get_last_bound(); return;
+        case PredicateCondition::GreaterThan: _end = _get_first_bound(); return;
+        default: Fail("Unsupported predicate condition encountered");
+      }
+    }
+    // clang-format on
+  }
+
+  /*
+   * NotEquals may result in two matching ranges (one below and one above the search_value) and needs special handling.
+   * The function contains three early outs. These are all only for performance reasons and, if removed, would not
+   * change the functionality.
+   */
+  template <typename Functor>
+  void _handle_not_equals(const Functor& functor) {
+    const auto first_occurrence = _get_first_bound();
+    if (first_occurrence == _end) {
+      // If the value isn't found anywhere, call the functor on the whole range and skip the call to _get_last_bound().
```

Comment:
Couldn't `first_occurrence` point to a value that is the first value before/behind the search value?

Reply:
Good catch! Although this didn't break the method, it allows for another early-out, which I added in 5e8f6cde481ed1a420cab15dd86d82ea39ac1c48. I also adjusted the comments to better reflect this possibility.